### PR TITLE
[Proposal] WASI-crypto Proposal

### PR DESCRIPTION
**Pull request by sonder-joker**
This is a long-term pull request that is still working, and will be constantly repair and commit for review. I think making this process public in early will work better.

### CODE REVIEW

File: utils/wasi-cpp-header/crypto_generate.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+ROOT="$(realpath $(dirname "${BASH_SOURCE[0]}")/../../)"
+cargo build --release
+
+for name in "asymmetric_common" "common" "external_secrets" "kx" "signatures" "symmetric" ;
+    do 
+        curl -s -L --remote-name-all https://raw.githubusercontent.com/WebAssembly/wasi-crypto/8e883a5a173062e1ecf7dbd7124188ab8046425d/witx/proposal_${name}.witx
+        # target/release/wasi-cpp-header generate --output "$ROOT/thirdparty/wasi/crypto/${name}.hpp" proposal_${name}.witx
+    done
+
+target/release/wasi-cpp-header generate --output /home/younger/WasmEdge/thirdparty/wasi/crypto/api.hpp proposal_kx.witx proposal_asymmetric_common.witx  proposal_common.witx proposal_signatures.witx proposal_symmetric.witx proposal_external_secrets.witx
```
Comment (hydai) [inline utils/wasi-cpp-header/crypto_generate.sh:14]: Don't use your local path in a general script.

**Comment by hydai** [comment]
Also, I think you will need a README file in the `docs/witx/crypto` which mentions that the version of `wasi-crypto`

**Comment by sonder-joker** [comment]
Only left many linux- I need check build script

File: include/common/enum_configure.h [hunk 2]
```diff
@@ -96,15 +96,25 @@ enum WasmEdge_Proposal {
 namespace WasmEdge {
 
 /// Host Module Registration C++ enumeration class.
-enum class HostRegistration : uint8_t { Wasi = 0, WasmEdge_Process, Max };
+enum class HostRegistration : uint8_t {
+  Wasi = 0,
+  WasmEdge_Process,
+#ifdef WASMEDGE_BUILD_WASI_CRYPTO
```
Comment (hydai) [inline include/common/enum_configure.h:102]: The HostRegistration should be a static one. So you should not use `ifdef` here to remove it. Instead, just keep it.

File: include/common/enum_configure.h [hunk 3]
```diff
@@ -96,15 +96,25 @@ enum WasmEdge_Proposal {
...
 
 } // namespace WasmEdge
 #endif
 
 /// Host Module Registration C enumeration.
 enum WasmEdge_HostRegistration {
   WasmEdge_HostRegistration_Wasi = 0,
-  WasmEdge_HostRegistration_WasmEdge_Process
+  WasmEdge_HostRegistration_WasmEdge_Process,
+#ifdef WASMEDGE_BUILD_WASI_CRYPTO
```
Comment (hydai) [inline include/common/enum_configure.h:115]: Ditto.

File: include/host/wasi_crypto/asymmetric_common/func.h [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  WasiCryptoContext &Ctx;
+};
+
+class KeypairGenerate : public HostFunction<KeypairGenerate> {
+public:
+  using HostFunction::HostFunction;
+
+  Expect<uint32_t> body(Runtime::Instance::MemoryInstance *MemInst,
+                        uint32_t AlgorithmType, const_uint8_t_ptr AlgorithmPtr,
+                        __wasi_size_t AlgorithmLen, uint32_t Options,
```
Comment (hydai) [inline include/host/wasi_crypto/asymmetric_common/func.h:29]: In the top level, we should use `stdint` like `int32_t`, `int64_t`, `uint32_t`, or `uint64_t`. Then, cast this type inside the function body.

E.g. `const_uint8_t_ptr` -> `uint32_t`; `__wasi_size_t` -> `uint32_t`.

File: include/host/wasi_crypto/asymmetric_common/func.h [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+                        uint8_t_ptr /* Out */ KeypairPtr);
```
Comment (hydai) [inline include/host/wasi_crypto/asymmetric_common/func.h:30]: Ditto.

File: lib/host/wasi_crypto/asymmetric_common/func.cpp [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+namespace Host {
+namespace WASICrypto {
+namespace AsymmetricCommon {
+
+Expect<uint32_t>
+KeypairGenerate::body(Runtime::Instance::MemoryInstance *MemInst,
+                      uint32_t AlgorithmType, const_uint8_t_ptr AlgorithmPtr,
+                      __wasi_size_t AlgorithmLen, uint32_t Options,
+                      uint8_t_ptr /* Out */ KeypairPtr) {
+  if (MemInst == nullptr) {
```
Comment (hydai) [inline lib/host/wasi_crypto/asymmetric_common/func.cpp:16]: Please use `unlikely(MemInst == nullptr)`

**Review by hydai — CHANGES_REQUESTED** [review]
There are some places which will need to fix first.

**Comment by sonder-joker** [comment]
Current this branch have a bad git log, so I make a new branch to refactor. This branch remain for reference.

**Pull request closed without merging**